File: src/com/train/algorithm/tree/implementInC++/Leaf-SimilarTrees.cpp
//
// Created by 罗斯 on 26/10/2022.
//
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaves1;
        vector<int> leaves2;
        getLeaves(root1, leaves1);
        getLeaves(root2, leaves2);
        return leaves1 == leaves2;
    }
private:
    void getLeaves(TreeNode* root, vector<int>& leaves){
        if (root == nullptr)
            return;
        if (root->left == nullptr && root->right == nullptr)
            leaves.push_back(root->val);
        getLeaves(root->left, leaves);
        getLeaves(root->right, leaves);
    }
};

File: src/com/train/algorithm/tree/implementInPython/Leaf-SimilarTrees.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def leafSimilar(self, root1, root2):
        def getLeaves(root):
            if not root: return []
            if not root.left and not root.right: return [root.val]
            return getLeaves(root.left) + getLeaves(root.right)
        return getLeaves(root1) == getLeaves(root2)